Source/ULibs.TinyJsonDeser/JsonDeserializer.cs

```diff
@@ -60,7 +60,7 @@ internal sealed class JsonDeserializer
         /// <c>bool</c>, a <c>string</c>, a <c>double</c>, an <c>object[]</c>, an <c>IDictionary&lt;string, object&gt;</c>
         /// or <c>null</c>.</param>
         /// <returns>Whether or not the parse attempt succeeded.</returns>
-        public bool TryParseValue(string json, out object output)
+        public bool TryParseValue(string json, out object? output)

```

Comment:
Apparently we can make these `Try*` methods slightly better by adding a `[NotNullWhen(returnValue: true)]` attribute to the out parameter? https://www.meziantou.net/csharp-8-nullable-reference-types.htm#conditional-post-con

Reply:
Hmm these attributes don't seem to be recognised

Reply:
Ah yeah, we might not have them here - sounds like we need to either update this project to `netstandard2.1` or write out the attributes manually? https://www.infoq.com/articles/CSharp-8-Framework/ . So long as we reference an attribute with the right name `System.Diagnostics.CodeAnalysis.NotNullWhenAttribute` I think it should work ...

(The latter option might be better since I think that means it'll work for `net472` as well?)

Reply:
So upgrading the projects to netstandard2.1 raises issues with the tests project (which is net4.6.1) and maybe other issues in dependencies?

In terms of referencing the attribute with the right name, this doesn't appear to work - those attributes don't exist on `System.Diagnostics.CodeAnalysis` in netstandard2.0 it appears. The work around in the article is to actually define the attributes manually which feels a bit overkill for this PR?

Reply:
> The work around in the article is to actually define the attributes manually which feels a bit overkill for this PR?

Sorry, yes, that's what I meant. I don't think it's too bad since we'd only actually have to define `NotNullWhenAttribute`, and it would make the `Try*` methods here more usable - it's not a huge deal though so I could go either way.

Reply:
Looks like that would work. However, it turns out that there are some occasions where output can be null and the return value can be true so I don't think this is a solution here. See line 197/198

Reply:
Ah, good spot. I think we need to look into this deeper - I'm guessing the new `if` statement around line 506 is probably a bug (since it'll filter out `null` values from the returned array?) but some of the others might still be necessary?

In `TryParseObjectKeyValuePair` I think we can add the attribute to `TryParseString`, which means that `key` should never be null, but then the value can totally be null so we'd need to change the type of the `KeyValuePair` appropriately.

It looks like there's a similar change in `SerializeGenericDictionary` which I'm now a bit worried about - I think we should be able to serialize a null value there too?

I'm a bit worried the build is still green at this point - we probably need to add some more unit tests.

Reply:
I don't think we should be changing any actual code here :/

Reply:
I agree - I am mostly trying to work out how it fits together...

Reply:
There's a fundamental problem here. You're trying to get this code to say it'll never return `null`, when that's a very explicitly valid case. Look at the XML doc comments. Line 61.